src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java

```diff
@@ -76,93 +92,166 @@ private void deleteWebhooks(List<BitbucketWebhook> webhooks) {
+                    supportedEvents.add(MIRROR_SYNCHRONIZED);
                 } else if (hooks.contains(REPO_REF_CHANGE.getEventId())) {
-                    return REPO_REF_CHANGE;
-                } else {
-                    throw new WebhookNotSupportedException("Remote server does not support the required events.");
+                    supportedEvents.add(REPO_REF_CHANGE);
                 }
-            } catch (BitbucketMissingCapabilityException exception) {
-                return REPO_REF_CHANGE;
+            } catch (BitbucketMissingCapabilityException exception) { //version doesn't support webhooks but support ref change & pr
+                    supportedEvents.add(REPO_REF_CHANGE);
+            }
+        } else {
+            if (request.isTriggerOnRefChange()) {
+                supportedEvents.add(REPO_REF_CHANGE);
+            }
+            if (request.isTriggerOnPullRequest()) {
+                supportedEvents.add(PULL_REQUEST_DECLINED);
+                supportedEvents.add(PULL_REQUEST_DELETED);
+                supportedEvents.add(PULL_REQUEST_FROM_REF_UPDATED);
+                supportedEvents.add(PULL_REQUEST_MERGED);
+                supportedEvents.add(PULL_REQUEST_OPENED);
             }
         }
-        return REPO_REF_CHANGE;
+        return supportedEvents;
     }
 
+    /**
+     * Returns the Bitbucket webhook we have just registered/updated.
+     * There should only be one webhook for each Jenkins to Bitbucket connection
+     * Possible webhook configurations: mirror sync, ref change, pull request events, ref change and pull request events
+     * 1. Registering new webhooks: This only happens when there is no mirror sync webhook and we need one, or if there are
+     * no webhooks listening for ref changes or pr events. If one exists, then we just update and add events to existing webhook.
+     * 2. Updating/deleting existing webhooks:
+     *  2.1 If our request has both ref change and pull request events: we must update any webhook with ref change or pull request events
+     *      and make sure both types of events are subscribed
+     *  2.2 If our request has only pull request events and we have an existing webhook with only pull request:
+     *      Don't change events subscribed (keep it as just pull request events)
+     *  2.3 If our request has only ref change and we have an existing webhook with only ref change:
+     *      Don't change events subscribed (keep it as just ref changes)
+     *  2.4 All other situations, subscribe to both ref changes to pull request events
+     *      e.g. If our request is only pull request events but our existing webhook is ref changes, we need our updated webhook to listen to both
+     * @param request the input request, collection of events
+     * @return the correct registered/updated bitbucket webhook
+     */
+    @Nullable
     private BitbucketWebhook process(WebhookRegisterRequest request,
```

Comment:
I think we really need to consider refactoring this method before we merge the PR. The existing system of just checking if we have matching triggers (which worked when there were only 2) is really straining now that we have 7, with four separate webhook lists we have to compare against. The lengthy conditional chains especially feel like an easy place for bugs to hide, and make detrimental changes accidentally.
I think this is complex enough to merit an object representation, and exposing methods like `hasWebhook`, so all these comparisons can be done internally. It would also mean we can add unit testing.